rpc/monero-rpc-types/src/lib.rs

```diff
@@ -1 +1,198 @@
+	// TODO
+	rustdoc::bare_urls,
+
+	clippy::module_name_repetitions,
+	clippy::module_inception,
+	clippy::redundant_pub_crate,
+	clippy::option_if_let_else,
+)]
+// Allow some lints when running in debug mode.
+#![cfg_attr(debug_assertions, allow(clippy::todo, clippy::multiple_crate_versions))]
+// Allow some lints in tests.
+#![cfg_attr(
+    test,
+    allow(
+        clippy::cognitive_complexity,
+        clippy::needless_pass_by_value,
+        clippy::cast_possible_truncation,
+        clippy::too_many_lines
+    )
+)]
+
+//---------------------------------------------------------------------------------------------------- Use
+// Misc types.
+mod status;
+pub use status::Status;
+
+// Internal modules.
+mod macros;
+
+// Request/response JSON/binary/other types.
+mod bin;
+mod json;
+mod other;
+
+/// TODO
+///
+/// TODO: explain
+/// - how this works
+/// - where to add types
+/// - when to add
+/// - what to do when adding/editing types
+macro_rules! re_export_request_and_response_types {
+    (
+		json {
+			$(
+				$json_type:ident,
+			)*
+		}
+		bin {
+			$(
+				$binary_type:ident,
+			)*
+		}
+		other {
+			$(
+				$other_type:ident,
+			)*
+		}
+	) => { paste::paste! {
+		/// RPC request types.
+		pub mod req {
+			/// JSON types from the [`/json_rpc`](https://www.getmonero.org/resources/developer-guides/daemon-rpc.html#json-rpc-methods) endpoint.
+			pub mod json {
+				$(
+					pub use $crate::json::[<Request $json_type>] as $json_type;
+				)*
+			}
+
+			/// Binary types from [binary](https://www.getmonero.org/resources/developer-guides/daemon-rpc.html#get_blocksbin) endpoints.
+			pub mod bin {
+				$(
+					pub use $crate::bin::[<Request $binary_type>] as $binary_type;
+				)*
+			}
+
+			/// JSON types from the [`other`](https://www.getmonero.org/resources/developer-guides/daemon-rpc.html#other-daemon-rpc-calls) endpoints.
+			pub mod other {
+				$(
+					pub use $crate::other::[<Request $other_type>] as $other_type;
+				)*
+			}
+		}
+
+		/// RPC response types.
+		pub mod resp {
+			/// JSON types from the [`/json_rpc`](https://www.getmonero.org/resources/developer-guides/daemon-rpc.html#json-rpc-methods) endpoint.
+			pub mod json {
+				$(
+					pub use $crate::json::[<Response $json_type>] as $json_type;
+				)*
+			}
+
+			/// Binary types from [binary](https://www.getmonero.org/resources/developer-guides/daemon-rpc.html#get_blocksbin) endpoints.
+			pub mod bin {
+				$(
+					pub use $crate::bin::[<Response $binary_type>] as $binary_type;
+				)*
+			}
+
+			/// JSON types from the [`other`](https://www.getmonero.org/resources/developer-guides/daemon-rpc.html#other-daemon-rpc-calls) endpoints.
+			pub mod other {
+				$(
+					pub use $crate::other::[<Response $other_type>] as $other_type;
+				)*
+			}
+		}
+	}};
+}
+
+re_export_request_and_response_types! {
+    json {
+        GetBlockCount,
+        OnGetBlockHash,
+        GetBlockTemplate,
+    }
+
+    bin {
+    }
+
+    other {
+        SaveBc,
+    }
+}
```

Comment:
The way the types/modules are setup are as follows:

- There's 3 actual files, `{bin,json,other}.rs`, which contain all the types with either a `Response/Request` suffix, e.g. `GetBlockCount{Request,Response}`
- This macro re-exports/re-routes/re-names those types to the appropriate module  e.g. `{req,resp}::json::GetBlockCount`

I'm not sure if there's a cleaner way to do this but I think modules are nicer than a `Response/Request` suffix on every type.

Reply:
People are often going to need the complimentary response and request type(s), which would have the same name so I don't know if this is a good idea, I think I would rather them just keep the suffix.

Reply:
Okay yeah now that I think about it the nested modules are probably not practical (pretty though).

How is this?:
```rust
use json_rpc::json::{GetBlockCountResponse, GetBlockCountRequest};
```
`req/resp` modules are removed, the top root modules are `{json,bin,other}`.

Reply:
Yeah that looks good